File: 백준/Gold/9663. N－Queen/N－Queen.cc
#include <iostream>
#include <vector>
#include <cmath> // abs() 함수 사용을 위해 포함

using namespace std;

int n, cnt = 0;
vector<int> board;
vector<bool> vis_col;

bool is_promising(int row) {
    for (int i = 0; i < row; ++i) {
        if (abs(i - row) == abs(board[i] - board[row])) {
            return false;
        }
    }
    return true;
}

void solution(int cur) {
    if (cur == n) {
        ++cnt;
        return;
    }

    for (int i = 0; i < n; ++i) {
        board[cur] = i;
        if (is_promising(cur) && !vis_col[i]) {
            vis_col[i] = true;
            solution(cur + 1);
            board[cur] = 0;
            vis_col[i] = false;
        }
    }
}

int main() {
    cin >> n;
    board.resize(n);
    vis_col.assign(n, false);

    solution(0);

    cout << cnt << endl;

    return 0;
}

File: 백준/Gold/9663. N－Queen/N－Queen.py
import sys

input = sys.stdin.readline


def solution(cur):
    if cur == n:
        global cnt
        cnt += 1
        return

    for i in range(n):
        if not flag_a[cur + i] and not flag_b[cur - i + n - 1] and not vis_col[i]:
            vis_col[i] = True
            flag_a[cur + i] = True
            flag_b[cur - i + n - 1] = True
            solution(cur + 1)
            vis_col[i] = False
            flag_a[cur + i] = False
            flag_b[cur - i + n - 1] = False


cnt = 0
n = int(input())
vis_col = [False] * n  # 세로
flag_a = [False] * n * 2  # 주대각선
flag_b = [False] * n * 2  # 부대각선
solution(0)
print(cnt)
